api/test/lib/middlewares/authentication_validator_test.dart

```diff
@@ -0,0 +1,156 @@
+
+class _MockContext extends Mock implements RequestContext {
+  @override
+  RequestContext provide<T extends Object?>(T Function() create) {
+    return this;
+  }
+}
+
+class _MockRequest extends Mock implements Request {}
+
+class _MockAuthenticationRepository extends Mock
+    implements AuthenticationRepository {}
+
+class _MockUserRepository extends Mock implements UserRepository {}
+
+void main() {
+  group('authenticationValidator', () {
+    late _MockContext context;
+    late _MockRequest request;
+    late _MockAuthenticationRepository authenticationRepository;
+    late _MockUserRepository userRepository;
+
+    setUp(() {
+      context = _MockContext();
+      request = _MockRequest();
+      authenticationRepository = _MockAuthenticationRepository();
+      userRepository = _MockUserRepository();
+
+      when(() => context.read<AuthenticationRepository>())
+          .thenReturn(authenticationRepository);
+      when(() => context.read<UserRepository>()).thenReturn(userRepository);
+
+      when(() => context.request).thenReturn(request);
+    });
+
+    test('returns the user when the session is valid', () async {
+      final session = Session(
+        id: 'id',
+        userId: 'userId',
+        token: '',
+        createdAt: DateTime.now(),
+        expiryDate: DateTime.now(),
+      );
+
+      const user = User(
+        id: 'userId',
+        username: 'username',
+        name: 'name',
+      );
+
+      when(() => authenticationRepository.verify('TOKEN'))
+          .thenReturn(session.toJson());
+
+      when(() => userRepository.findUserById('userId')).thenAnswer(
+        (_) async => user,
+      );
+
+      when(() => request.headers).thenReturn(
+        {
+          'Authorization': 'Bearer TOKEN',
+        },
+      );
+
+      final middleware = authenticationValidator();
+
+      var called = false;
+      await middleware((_) {
+        called = true;
+        return Response();
+      })(context);
+
+      expect(called, isTrue);
+    });
+
+    test('returns null when the is no authorization token', () async {
+      when(() => request.headers).thenReturn(
+        {},
+      );
+
+      final middleware = authenticationValidator();
+
+      var called = false;
+      await middleware((_) {
+        called = true;
+        return Response();
+      })(context);
+
+      expect(called, isFalse);
+    });
+
+    test('returns null when the token is not valid', () async {
+      when(() => authenticationRepository.verify('TOKEN'))
+          .thenThrow(Exception());
+
+      when(() => request.headers).thenReturn(
+        {
+          'Authorization': 'Bearer TOKEN',
+        },
+      );
+
+      final middleware = authenticationValidator();
+
+      var called = false;
+      try {
+        await middleware((_) {
+          called = true;
+          return Response();
+        })(context);
+      } catch (_) {
+        // Expected
+      }
+
+      expect(called, isFalse);
```

Comment:
Maybe we could replace the try/catch by doing something like this:
```suggestion
      final middleware = authenticationValidator();

      expect(
        middleware((_) => Response())(context),
        throwsA(isA<Exception>()),
      );
```

Reply:
We could, but I feel that we wouldn't be really testing, because we know for sure that an exception will be throw, since we mocked that on the above lines.

But more than that, we want to be sure that the closure is not called when something goes wrong. If we change to assert with a throwsA, we would not really be asserting what this test should be verifying. Makes sense?

Reply:
Yes, that makes sense. Maybe the description of the test could be updated to match more closely what is being testing.

Reply:
Fair enough!!! I have update the description to a better description!